bug: invalid recursive type

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version of ogen are you using?

```bash
$ go list -m github.com/ogen-go/ogen
```

github.com/ogen-go/ogen v1.2.1

### Can this issue be reproduced with the latest version?

Yes

### What did you do?

Generate spec 
https://github.com/gotd/botapi/blob/6ca9948cfcdc94eed4ff53dd364672ac791a9db3/_oas/openapi.json

### What did you expect to see?

Generated code compiles

### What did you see instead?

```
ernado@work:/src/gotd/botapi$ make test 
?       github.com/gotd/botapi  [no test files]
# github.com/gotd/botapi/internal/oas
internal/oas/oas_schemas_gen.go:7566:6: invalid recursive type GiveawayCompleted
        internal/oas/oas_schemas_gen.go:7566:6: GiveawayCompleted refers to
        internal/oas/oas_schemas_gen.go:19031:6: OptMessage refers to
        internal/oas/oas_schemas_gen.go:14271:6: Message refers to
        internal/oas/oas_schemas_gen.go:18065:6: OptGiveawayCompleted refers to
        internal/oas/oas_schemas_gen.go:7566:6: GiveawayCompleted
internal/oas/oas_json_gen.go:32550:11: s.MessageID undefined (type *Message has no field or method MessageID)
internal/oas/oas_json_gen.go:32553:8: s.MessageThreadID undefined (type *Message has no field or method MessageThreadID)
internal/oas/oas_json_gen.go:32555:6: s.MessageThreadID undefined (type *Message has no field or method MessageThreadID)
internal/oas/oas_json_gen.go:32559:8: s.From undefined (type *Message has no field or method From)
internal/oas/oas_json_gen.go:32561:6: s.From undefined (type *Message has no field or method From)
internal/oas/oas_json_gen.go:32565:8: s.SenderChat undefined (type *Message has no field or method SenderChat)
internal/oas/oas_json_gen.go:32567:6: s.SenderChat undefined (type *Message has no field or method SenderChat)
internal/oas/oas_json_gen.go:32571:8: s.SenderBoostCount undefined (type *Message has no field or method SenderBoostCount)
internal/oas/oas_json_gen.go:32573:6: s.SenderBoostCount undefined (type *Message has no field or method SenderBoostCount)
internal/oas/oas_json_gen.go:32573:6: too many errors
```

--- a/examples/ex_gotd/oas_faker_gen.go
+++ b/examples/ex_gotd/oas_faker_gen.go
@@ -454,8 +454,8 @@ func (s *CallbackQuery) SetFake() {
 		}
 	}
 	{
-		{
-			s.Message.SetFake()
+		{ // Keep pointer nil to prevent infinite recursion.
+			s.Message = nil
 		}
 	}
 	{

--- a/examples/ex_gotd/oas_json_gen.go
+++ b/examples/ex_gotd/oas_json_gen.go
@@ -2893,7 +2893,7 @@ func (s *CallbackQuery) encodeFields(e *jx.Encoder) {
 		s.From.Encode(e)
 	}
 	{
-		if s.Message.Set {
+		if s.Message != nil {
 			e.FieldStart("message")
 			s.Message.Encode(e)
 		}
@@ -2965,10 +2965,12 @@ func (s *CallbackQuery) Decode(d *jx.Decoder) error {
 			}
 		case "message":
 			if err := func() error {
-				s.Message.Reset()
-				if err := s.Message.Decode(d); err != nil {
+				s.Message = nil
+				var elem Message
+				if err := elem.Decode(d); err != nil {
 					return err
 				}
+				s.Message = &elem
 				return nil
 			}(); err != nil {
 				return errors.Wrap(err, "decode field \"message\"")

--- a/examples/ex_gotd/oas_schemas_gen.go
+++ b/examples/ex_gotd/oas_schemas_gen.go
@@ -1161,9 +1161,9 @@ type CallbackGame struct{}
 // Ref: #/components/schemas/CallbackQuery
 type CallbackQuery struct {
 	// Unique identifier for this query.
-	ID      string     `json:"id"`
-	From    User       `json:"from"`
-	Message OptMessage `json:"message"`
+	ID      string   `json:"id"`
+	From    User     `json:"from"`
+	Message *Message `json:"message"`
 	// Identifier of the message sent via the bot in inline mode, that originated the query.
 	InlineMessageID OptString `json:"inline_message_id"`
 	// Global identifier, uniquely corresponding to the chat to which the message with the callback
@@ -1187,7 +1187,7 @@ func (s *CallbackQuery) GetFrom() User {
 }
 
 // GetMessage returns the value of Message.
-func (s *CallbackQuery) GetMessage() OptMessage {
+func (s *CallbackQuery) GetMessage() *Message {
 	return s.Message
 }
 
@@ -1222,7 +1222,7 @@ func (s *CallbackQuery) SetFrom(val User) {
 }
 
 // SetMessage sets the value of Message.
-func (s *CallbackQuery) SetMessage(val OptMessage) {
+func (s *CallbackQuery) SetMessage(val *Message) {
 	s.Message = val
 }
 

--- a/examples/ex_gotd/oas_validators_gen.go
+++ b/examples/ex_gotd/oas_validators_gen.go
@@ -401,15 +401,16 @@ func (s *CallbackQuery) Validate() error {
 
 	var failures []validate.FieldError
 	if err := func() error {
-		if value, ok := s.Message.Get(); ok {
-			if err := func() error {
-				if err := value.Validate(); err != nil {
-					return err
-				}
-				return nil
-			}(); err != nil {
+		if s.Message == nil {
+			return nil // optional
+		}
+		if err := func() error {
+			if err := s.Message.Validate(); err != nil {
 				return err
 			}
+			return nil
+		}(); err != nil {
+			return errors.Wrap(err, "pointer")
 		}
 		return nil
 	}(); err != nil {

--- a/examples/ex_k8s/oas_faker_gen.go
+++ b/examples/ex_k8s/oas_faker_gen.go
@@ -14276,8 +14276,8 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefiniti
 		}
 	}
 	{
-		{
-			s.Schema.SetFake()
+		{ // Keep pointer nil to prevent infinite recursion.
+			s.Schema = nil
 		}
 	}
 	{
@@ -14337,8 +14337,8 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceSubresou
 // SetFake set fake values.
 func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation) SetFake() {
 	{
-		{
-			s.OpenAPIV3Schema.SetFake()
+		{ // Keep pointer nil to prevent infinite recursion.
+			s.OpenAPIV3Schema = nil
 		}
 	}
 }

--- a/examples/ex_k8s/oas_json_gen.go
+++ b/examples/ex_k8s/oas_json_gen.go
@@ -70139,7 +70139,7 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefiniti
 		e.Str(s.Name)
 	}
 	{
-		if s.Schema.Set {
+		if s.Schema != nil {
 			e.FieldStart("schema")
 			s.Schema.Encode(e)
 		}
@@ -70231,10 +70231,12 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefiniti
 			}
 		case "schema":
 			if err := func() error {
-				s.Schema.Reset()
-				if err := s.Schema.Decode(d); err != nil {
+				s.Schema = nil
+				var elem IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation
+				if err := elem.Decode(d); err != nil {
 					return err
 				}
+				s.Schema = &elem
 				return nil
 			}(); err != nil {
 				return errors.Wrap(err, "decode field \"schema\"")
@@ -70595,7 +70597,7 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidati
 // encodeFields encodes fields.
 func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation) encodeFields(e *jx.Encoder) {
 	{
-		if s.OpenAPIV3Schema.Set {
+		if s.OpenAPIV3Schema != nil {
 			e.FieldStart("openAPIV3Schema")
 			s.OpenAPIV3Schema.Encode(e)
 		}
@@ -70616,10 +70618,12 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidati
 		switch string(k) {
 		case "openAPIV3Schema":
 			if err := func() error {
-				s.OpenAPIV3Schema.Reset()
-				if err := s.OpenAPIV3Schema.Decode(d); err != nil {
+				s.OpenAPIV3Schema = nil
+				var elem IoK8sApiextensionsApiserverPkgApisApiextensionsV1JSONSchemaProps
+				if err := elem.Decode(d); err != nil {
 					return err
 				}
+				s.OpenAPIV3Schema = &elem
 				return nil
 			}(); err != nil {
 				return errors.Wrap(err, "decode field \"openAPIV3Schema\"")

--- a/examples/ex_k8s/oas_schemas_gen.go
+++ b/examples/ex_k8s/oas_schemas_gen.go
@@ -31776,8 +31776,8 @@ type IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefinitionVe
 	DeprecationWarning OptString `json:"deprecationWarning"`
 	// Name is the version name, e.g. “v1”, “v2beta1”, etc. The custom resources are served under
 	// this version at `/apis/<group>/<version>/...` if `served` is true.
-	Name   string                                                                       `json:"name"`
-	Schema OptIoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation `json:"schema"`
+	Name   string                                                                     `json:"name"`
+	Schema *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation `json:"schema"`
 	// Served is a flag enabling/disabling this version from being served via REST APIs.
 	Served bool `json:"served"`
 	// Storage indicates this version should be used when persisting custom resources to storage. There
@@ -31807,7 +31807,7 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefiniti
 }
 
 // GetSchema returns the value of Schema.
-func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefinitionVersion) GetSchema() OptIoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation {
+func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefinitionVersion) GetSchema() *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation {
 	return s.Schema
 }
 
@@ -31847,7 +31847,7 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefiniti
 }
 
 // SetSchema sets the value of Schema.
-func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefinitionVersion) SetSchema(val OptIoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation) {
+func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefinitionVersion) SetSchema(val *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation) {
 	s.Schema = val
 }
 
@@ -31958,16 +31958,16 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceSubresou
 // CustomResourceValidation is a list of validation methods for CustomResources.
 // Ref: #/components/schemas/io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceValidation
 type IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation struct {
-	OpenAPIV3Schema OptIoK8sApiextensionsApiserverPkgApisApiextensionsV1JSONSchemaProps `json:"openAPIV3Schema"`
+	OpenAPIV3Schema *IoK8sApiextensionsApiserverPkgApisApiextensionsV1JSONSchemaProps `json:"openAPIV3Schema"`
 }
 
 // GetOpenAPIV3Schema returns the value of OpenAPIV3Schema.
-func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation) GetOpenAPIV3Schema() OptIoK8sApiextensionsApiserverPkgApisApiextensionsV1JSONSchemaProps {
+func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation) GetOpenAPIV3Schema() *IoK8sApiextensionsApiserverPkgApisApiextensionsV1JSONSchemaProps {
 	return s.OpenAPIV3Schema
 }
 
 // SetOpenAPIV3Schema sets the value of OpenAPIV3Schema.
-func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation) SetOpenAPIV3Schema(val OptIoK8sApiextensionsApiserverPkgApisApiextensionsV1JSONSchemaProps) {
+func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidation) SetOpenAPIV3Schema(val *IoK8sApiextensionsApiserverPkgApisApiextensionsV1JSONSchemaProps) {
 	s.OpenAPIV3Schema = val
 }
 

--- a/examples/ex_k8s/oas_validators_gen.go
+++ b/examples/ex_k8s/oas_validators_gen.go
@@ -4739,15 +4739,16 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceDefiniti
 
 	var failures []validate.FieldError
 	if err := func() error {
-		if value, ok := s.Schema.Get(); ok {
-			if err := func() error {
-				if err := value.Validate(); err != nil {
-					return err
-				}
-				return nil
-			}(); err != nil {
+		if s.Schema == nil {
+			return nil // optional
+		}
+		if err := func() error {
+			if err := s.Schema.Validate(); err != nil {
 				return err
 			}
+			return nil
+		}(); err != nil {
+			return errors.Wrap(err, "pointer")
 		}
 		return nil
 	}(); err != nil {
@@ -4769,15 +4770,16 @@ func (s *IoK8sApiextensionsApiserverPkgApisApiextensionsV1CustomResourceValidati
 
 	var failures []validate.FieldError
 	if err := func() error {
-		if value, ok := s.OpenAPIV3Schema.Get(); ok {
-			if err := func() error {
-				if err := value.Validate(); err != nil {
-					return err
-				}
-				return nil
-			}(); err != nil {
+		if s.OpenAPIV3Schema == nil {
+			return nil // optional
+		}
+		if err := func() error {
+			if err := s.OpenAPIV3Schema.Validate(); err != nil {
 				return err
 			}
+			return nil
+		}(); err != nil {
+			return errors.Wrap(err, "pointer")
 		}
 		return nil
 	}(); err != nil {

--- a/examples/ex_openapi/output.gen.go
+++ b/examples/ex_openapi/output.gen.go
@@ -702,7 +702,7 @@ type Header struct {
 	Style           OptString               `json:"style"`
 	Explode         OptBool                 `json:"explode"`
 	AllowReserved   OptBool                 `json:"allowReserved"`
-	Schema          OptSchemaOrReference    `json:"schema"`
+	Schema          *SchemaOrReference      `json:"schema"`
 	Example         Any                     `json:"example"`
 	Examples        OptExamplesOrReferences `json:"examples"`
 	Content         OptMediaTypes           `json:"content"`
@@ -746,7 +746,7 @@ func (s *Header) GetAllowReserved() OptBool {
 }
 
 // GetSchema returns the value of Schema.
-func (s *Header) GetSchema() OptSchemaOrReference {
+func (s *Header) GetSchema() *SchemaOrReference {
 	return s.Schema
 }
 
@@ -806,7 +806,7 @@ func (s *Header) SetAllowReserved(val OptBool) {
 }
 
 // SetSchema sets the value of Schema.
-func (s *Header) SetSchema(val OptSchemaOrReference) {
+func (s *Header) SetSchema(val *SchemaOrReference) {
 	s.Schema = val
 }
 
@@ -1245,7 +1245,7 @@ type Maximum float64
 // Each Media Type Object provides schema and examples for the media type identified by its key.
 // Ref: #/definitions/mediaType
 type MediaType struct {
-	Schema   OptSchemaOrReference    `json:"schema"`
+	Schema   *SchemaOrReference      `json:"schema"`
 	Example  Any                     `json:"example"`
 	Examples OptExamplesOrReferences `json:"examples"`
 	Encoding OptEncodings            `json:"encoding"`
@@ -1254,7 +1254,7 @@ type MediaType struct {
 }
 
 // GetSchema returns the value of Schema.
-func (s *MediaType) GetSchema() OptSchemaOrReference {
+func (s *MediaType) GetSchema() *SchemaOrReference {
 	return s.Schema
 }
 
@@ -1279,7 +1279,7 @@ func (s *MediaType) GetPattern0Props() MediaTypePattern0 {
 }
 
 // SetSchema sets the value of Schema.
-func (s *MediaType) SetSchema(val OptSchemaOrReference) {
+func (s *MediaType) SetSchema(val *SchemaOrReference) {
 	s.Schema = val
 }
 
@@ -3625,7 +3625,7 @@ type Parameter struct {
 	Style           OptParameterStyle       `json:"style"`
 	Explode         OptBool                 `json:"explode"`
 	AllowReserved   OptBool                 `json:"allowReserved"`
-	Schema          OptSchemaOrReference    `json:"schema"`
+	Schema          *SchemaOrReference      `json:"schema"`
 	Example         Any                     `json:"example"`
 	Examples        OptExamplesOrReferences `json:"examples"`
 	Content         OptMediaTypes           `json:"content"`
@@ -3679,7 +3679,7 @@ func (s *Parameter) GetAllowReserved() OptBool {
 }
 
 // GetSchema returns the value of Schema.
-func (s *Parameter) GetSchema() OptSchemaOrReference {
+func (s *Parameter) GetSchema() *SchemaOrReference {
 	return s.Schema
 }
 
@@ -3749,7 +3749,7 @@ func (s *Parameter) SetAllowReserved(val OptBool) {
 }
 
 // SetSchema sets the value of Schema.
-func (s *Parameter) SetSchema(val OptSchemaOrReference) {
+func (s *Parameter) SetSchema(val *SchemaOrReference) {
 	s.Schema = val
 }
 
@@ -7995,7 +7995,7 @@ func (s *Header) encodeFields(e *jx.Encoder) {
 		}
 	}
 	{
-		if s.Schema.Set {
+		if s.Schema != nil {
 			e.FieldStart("schema")
 			s.Schema.Encode(e)
 		}
@@ -8118,10 +8118,12 @@ func (s *Header) Decode(d *jx.Decoder) error {
 			}
 		case "schema":
 			if err := func() error {
-				s.Schema.Reset()
-				if err := s.Schema.Decode(d); err != nil {
+				s.Schema = nil
+				var elem SchemaOrReference
+				if err := elem.Decode(d); err != nil {
 					return err
 				}
+				s.Schema = &elem
 				return nil
 			}(); err != nil {
 				return errors.Wrap(err, "decode field \"schema\"")
@@ -9410,7 +9412,7 @@ func (s *MediaType) Encode(e *jx.Encoder) {
 // encodeFields encodes fields.
 func (s *MediaType) encodeFields(e *jx.Encoder) {
 	{
-		if s.Schema.Set {
+		if s.Schema != nil {
 			e.FieldStart("schema")
 			s.Schema.Encode(e)
 		}
@@ -9456,10 +9458,12 @@ func (s *MediaType) Decode(d *jx.Decoder) error {
 		switch string(k) {
 		case "schema":
 			if err := func() error {
-				s.Schema.Reset()
-				if err := s.Schema.Decode(d); err != nil {
+				s.Schema = nil
+				var elem SchemaOrReference
+				if err := elem.Decode(d); err != nil {
 					return err
 				}
+				s.Schema = &elem
 				return nil
 			}(); err != nil {
 				return errors.Wrap(err, "decode field \"schema\"")
@@ -12047,7 +12051,7 @@ func (s *Parameter) encodeFields(e *jx.Encoder) {
 		}
 	}
 	{
-		if s.Schema.Set {
+		if s.Schema != nil {
 			e.FieldStart("schema")
 			s.Schema.Encode(e)
 		}
@@ -12195,10 +12199,12 @@ func (s *Parameter) Decode(d *jx.Decoder) error {
 			}
 		case "schema":
 			if err := func() error {
-				s.Schema.Reset()
-				if err := s.Schema.Decode(d); err != nil {
+				s.Schema = nil
+				var elem SchemaOrReference
+				if err := elem.Decode(d); err != nil {
 					return err
 				}
+				s.Schema = &elem
 				return nil
 			}(); err != nil {
 				return errors.Wrap(err, "decode field \"schema\"")
@@ -18143,15 +18149,16 @@ func (s *Header) Validate() error {
 
 	var failures []validate.FieldError
 	if err := func() error {
-		if value, ok := s.Schema.Get(); ok {
-			if err := func() error {
-				if err := value.Validate(); err != nil {
-					return err
-				}
-				return nil
-			}(); err != nil {
+		if s.Schema == nil {
+			return nil // optional
+		}
+		if err := func() error {
+			if err := s.Schema.Validate(); err != nil {
 				return err
 			}
+			return nil
+		}(); err != nil {
+			return errors.Wrap(err, "pointer")
 		}
 		return nil
 	}(); err != nil {
@@ -18235,15 +18242,16 @@ func (s *MediaType) Validate() error {
 
 	var failures []validate.FieldError
 	if err := func() error {
-		if value, ok := s.Schema.Get(); ok {
-			if err := func() error {
-				if err := value.Validate(); err != nil {
-					return err
-				}
-				return nil
-			}(); err != nil {
+		if s.Schema == nil {
+			return nil // optional
+		}
+		if err := func() error {
+			if err := s.Schema.Validate(); err != nil {
 				return err
 			}
+			return nil
+		}(); err != nil {
+			return errors.Wrap(err, "pointer")
 		}
 		return nil
 	}(); err != nil {
@@ -18468,15 +18476,16 @@ func (s *Parameter) Validate() error {
 		})
 	}
 	if err := func() error {
-		if value, ok := s.Schema.Get(); ok {
-			if err := func() error {
-				if err := value.Validate(); err != nil {
-					return err
-				}
-				return nil
-			}(); err != nil {
+		if s.Schema == nil {
+			return nil // optional
+		}
+		if err := func() error {
+			if err := s.Schema.Validate(); err != nil {
 				return err
 			}
+			return nil
+		}(); err != nil {
+			return errors.Wrap(err, "pointer")
 		}
 		return nil
 	}(); err != nil {

--- a/examples/ex_telegram/oas_faker_gen.go
+++ b/examples/ex_telegram/oas_faker_gen.go
@@ -439,8 +439,8 @@ func (s *CallbackQuery) SetFake() {
 		}
 	}
 	{
-		{
-			s.Message.SetFake()
+		{ // Keep pointer nil to prevent infinite recursion.
+			s.Message = nil
 		}
 	}
 	{

--- a/examples/ex_telegram/oas_json_gen.go
+++ b/examples/ex_telegram/oas_json_gen.go
@@ -2766,7 +2766,7 @@ func (s *CallbackQuery) encodeFields(e *jx.Encoder) {
 		s.From.Encode(e)
 	}
 	{
-		if s.Message.Set {
+		if s.Message != nil {
 			e.FieldStart("message")
 			s.Message.Encode(e)
 		}
@@ -2838,10 +2838,12 @@ func (s *CallbackQuery) Decode(d *jx.Decoder) error {
 			}
 		case "message":
 			if err := func() error {
-				s.Message.Reset()
-				if err := s.Message.Decode(d); err != nil {
+				s.Message = nil
+				var elem Message
+				if err := elem.Decode(d); err != nil {
 					return err
 				}
+				s.Message = &elem
 				return nil
 			}(); err != nil {
 				return errors.Wrap(err, "decode field \"message\"")

--- a/examples/ex_telegram/oas_schemas_gen.go
+++ b/examples/ex_telegram/oas_schemas_gen.go
@@ -1121,9 +1121,9 @@ type CallbackGame struct{}
 // Ref: #/components/schemas/CallbackQuery
 type CallbackQuery struct {
 	// Unique identifier for this query.
-	ID      string     `json:"id"`
-	From    User       `json:"from"`
-	Message OptMessage `json:"message"`
+	ID      string   `json:"id"`
+	From    User     `json:"from"`
+	Message *Message `json:"message"`
 	// Identifier of the message sent via the bot in inline mode, that originated the query.
 	InlineMessageID OptString `json:"inline_message_id"`
 	// Global identifier, uniquely corresponding to the chat to which the message with the callback
@@ -1147,7 +1147,7 @@ func (s *CallbackQuery) GetFrom() User {
 }
 
 // GetMessage returns the value of Message.
-func (s *CallbackQuery) GetMessage() OptMessage {
+func (s *CallbackQuery) GetMessage() *Message {
 	return s.Message
 }
 
@@ -1182,7 +1182,7 @@ func (s *CallbackQuery) SetFrom(val User) {
 }
 
 // SetMessage sets the value of Message.
-func (s *CallbackQuery) SetMessage(val OptMessage) {
+func (s *CallbackQuery) SetMessage(val *Message) {
 	s.Message = val
 }
 

--- a/examples/ex_telegram/oas_validators_gen.go
+++ b/examples/ex_telegram/oas_validators_gen.go
@@ -378,15 +378,16 @@ func (s *CallbackQuery) Validate() error {
 
 	var failures []validate.FieldError
 	if err := func() error {
-		if value, ok := s.Message.Get(); ok {
-			if err := func() error {
-				if err := value.Validate(); err != nil {
-					return err
-				}
-				return nil
-			}(); err != nil {
+		if s.Message == nil {
+			return nil // optional
+		}
+		if err := func() error {
+			if err := s.Message.Validate(); err != nil {
 				return err
 			}
+			return nil
+		}(); err != nil {
+			return errors.Wrap(err, "pointer")
 		}
 		return nil
 	}(); err != nil {

--- a/gen/gen_schema.go
+++ b/gen/gen_schema.go
@@ -8,28 +8,18 @@ import (
 	"go.uber.org/zap"
 
 	"github.com/ogen-go/ogen/gen/ir"
+	"github.com/ogen-go/ogen/internal/xmaps"
 	"github.com/ogen-go/ogen/jsonschema"
 )
 
 func saveSchemaTypes(ctx *genctx, gen *schemaGen, refEncoding map[jsonschema.Ref]ir.Encoding) error {
 	for _, t := range gen.side {
-		if t.IsStruct() {
-			if err := checkStructRecursions(ctx, t); err != nil {
-				return errors.Wrap(err, t.Name)
-			}
-		}
-
 		if err := ctx.saveType(t); err != nil {
 			return errors.Wrap(err, "save inlined type")
 		}
 	}
 
 	for ref, t := range gen.localRefs {
-		if t.IsStruct() {
-			if err := checkStructRecursions(ctx, t); err != nil {
-				return errors.Wrap(err, ref.String())
-			}
-		}
 		encoding := ir.EncodingJSON
 		if e, ok := refEncoding[ref]; ok {
 			encoding = e
@@ -177,6 +167,15 @@ func GenerateSchema(schema *jsonschema.Schema, fs FileSystem, opts GenerateSchem
 
 	if err := saveSchemaTypes(ctx, gen, nil); err != nil {
 		return errors.Wrap(err, "save schema types")
+	}
+
+	types := ctx.local.types
+	for _, key := range xmaps.SortedKeys(types) {
+		if t := types[key]; t.IsStruct() {
+			if err := checkStructRecursions(t); err != nil {
+				return errors.Wrap(err, t.Name)
+			}
+		}
 	}
 
 	w := &writer{

--- a/gen/generator.go
+++ b/gen/generator.go
@@ -13,6 +13,7 @@ import (
 
 	"github.com/ogen-go/ogen"
 	"github.com/ogen-go/ogen/gen/ir"
+	"github.com/ogen-go/ogen/internal/xmaps"
 	"github.com/ogen-go/ogen/internal/xslices"
 	"github.com/ogen-go/ogen/jsonschema"
 	"github.com/ogen-go/ogen/openapi"
@@ -187,6 +188,15 @@ func (g *Generator) makeOps(ops []*openapi.Operation) error {
 		}
 
 		g.operations = append(g.operations, op)
+	}
+
+	types := g.Types()
+	for _, key := range xmaps.SortedKeys(types) {
+		if t := types[key]; t.IsStruct() {
+			if err := checkStructRecursions(t); err != nil {
+				return errors.Wrap(err, t.Name)
+			}
+		}
 	}
 
 	sortOperations(g.operations)

--- a/gen/generics.go
+++ b/gen/generics.go
@@ -7,7 +7,7 @@ import (
 	"github.com/ogen-go/ogen/internal/naming"
 )
 
-func checkStructRecursions(ctx *genctx, s *ir.Type) error {
+func checkStructRecursions(s *ir.Type) error {
 	for _, field := range s.Fields {
 		if field.Spec == nil {
 			continue
@@ -36,11 +36,6 @@ func checkStructRecursions(ctx *genctx, s *ir.Type) error {
 					if err != nil {
 						return nil, err
 					}
-
-					if err := ctx.saveType(t); err != nil {
-						return nil, err
-					}
-
 					return ir.Pointer(t, ir.NilNull), nil
 				default:
 					// Required.

--- a/gen/ir/recursion.go
+++ b/gen/ir/recursion.go
@@ -27,14 +27,14 @@ func (t *Type) recursive(target *Type, path *walkpath) bool {
 		}
 	}
 
-	if reflect.DeepEqual(t, target) {
+	if t == target || reflect.DeepEqual(t, target) {
 		return true
 	}
-
 	if path.has(target) {
-		return false
+		return true
 	}
-	path = path.append(target)
+	path.add(target)
+	defer path.delete(target)
 
 	switch target.Kind {
 	case KindAlias:
@@ -42,11 +42,7 @@ func (t *Type) recursive(target *Type, path *walkpath) bool {
 	case KindGeneric:
 		return t.recursive(target.GenericOf, path)
 	case KindStruct:
-		return slices.ContainsFunc(target.Fields, func(f *Field) bool {
-			// Ignore optional fields: we are using pointers for them.
-			if f.Spec != nil && !f.Spec.Required {
-				return false
-			}
+		return slices.ContainsFunc(target.Fields, func(f *Field) (r bool) {
 			return t.recursive(f.Type, path)
 		})
 	case KindSum:
@@ -59,18 +55,21 @@ func (t *Type) recursive(target *Type, path *walkpath) bool {
 }
 
 type walkpath struct {
-	nodes []*Type
+	nodes map[*Type]struct{}
 }
 
 func (wp *walkpath) has(t *Type) bool {
-	return slices.Contains(wp.nodes, t)
+	_, ok := wp.nodes[t]
+	return ok
 }
 
-func (wp *walkpath) append(t *Type) *walkpath {
-	return &walkpath{
-		append(
-			wp.nodes[:len(wp.nodes):len(wp.nodes)],
-			t,
-		),
+func (wp *walkpath) add(t *Type) {
+	if wp.nodes == nil {
+		wp.nodes = map[*Type]struct{}{}
 	}
+	wp.nodes[t] = struct{}{}
+}
+
+func (wp *walkpath) delete(t *Type) {
+	delete(wp.nodes, t)
 }

--- a/gen/ir/validation.go
+++ b/gen/ir/validation.go
@@ -136,8 +136,8 @@ func (t *Type) needValidation(path *walkpath) (result bool) {
 	if path.has(t) {
 		return false
 	}
-
-	path = path.append(t)
+	path.add(t)
+	defer path.delete(t)
 
 	switch t.Kind {
 	case KindPrimitive:

--- a/internal/integration/test_naming_extensions/oas_json_gen.go
+++ b/internal/integration/test_naming_extensions/oas_json_gen.go
@@ -12,6 +12,39 @@ import (
 	"github.com/ogen-go/ogen/validate"
 )
 
+// Encode encodes Person as json.
+func (o OptPerson) Encode(e *jx.Encoder) {
+	if !o.Set {
+		return
+	}
+	o.Value.Encode(e)
+}
+
+// Decode decodes Person from json.
+func (o *OptPerson) Decode(d *jx.Decoder) error {
+	if o == nil {
+		return errors.New("invalid: unable to decode OptPerson to nil")
+	}
+	o.Set = true
+	if err := o.Value.Decode(d); err != nil {
+		return err
+	}
+	return nil
+}
+
+// MarshalJSON implements stdjson.Marshaler.
+func (s OptPerson) MarshalJSON() ([]byte, error) {
+	e := jx.Encoder{}
+	s.Encode(&e)
+	return e.Bytes(), nil
+}
+
+// UnmarshalJSON implements stdjson.Unmarshaler.
+func (s *OptPerson) UnmarshalJSON(data []byte) error {
+	d := jx.DecodeBytes(data)
+	return s.Decode(d)
+}
+
 // Encode implements json.Marshaler.
 func (s *Person) Encode(e *jx.Encoder) {
 	e.ObjStart()

